### PR TITLE
[Fix] get textStart from pclnTable

### PR DESCRIPTION
**Pull request by wlingze**
fixup: #512
details： wlingze#2， #514

### CODE REVIEW

File: user/config/config_gotls.go [hunk 1]
```diff
@@ -311,8 +305,18 @@ func (gc *GoTLSConfig) ReadTable() (*gosym.Table, error) {
...
 	}
 	tableData = tableData[pclntabIndex:]
-
-	addr := gc.goElf.Section(".text").Addr
+	var addr uint64
+	{
+		// get textStart from pclntable
+		// please see https://go-review.googlesource.com/c/go/+/366695
+		// tableData
+		ptrSize := uint32(tableData[7])
```
Comment (cfc4n) [inline user/config/config_gotls.go:313]: Is 7 a fixed index position?

File: user/config/config_gotls.go [hunk 2]
```diff
@@ -311,8 +305,18 @@ func (gc *GoTLSConfig) ReadTable() (*gosym.Table, error) {
...
+		if ptrSize == 4 {
+			addr = uint64(binary.LittleEndian.Uint32(tableData[8+2*ptrSize:]))
+		} else {
+			addr = binary.LittleEndian.Uint64(tableData[8+2*ptrSize:])
```
Comment (cfc4n) [inline user/config/config_gotls.go:317]: What is the relationship between `gc.goElf.Section(".text").Addr `  and this ?

File: user/config/config_gotls.go (hunk 1 above)
Comment (wlingze) [reply user/config/config_gotls.go:313]: yes, please see [pclntab.go#L249](https://github.com/golang/go/blob/master/src/debug/gosym/pclntab.go#L249), ` ptrSize are the same between 1.2, 1.16, and 1.18`

File: user/config/config_gotls.go (hunk 2 above)
Comment (wlingze) [reply user/config/config_gotls.go:317]: Well, I don't know the relationship between `.text` section address and this address ("runtime.text"), I think in some cases, there are same address, but no more. I just know relationship between"runtime.text" address and function address.

Package `debug/gosym` implement parse `pclntab`, in this code, we'll see [go pr 366695](https://go-review.googlesource.com/c/go/+/366695), this commit show `textStart` should be got from `pclntab` not address of `.text` section, and this address only use in golang version >= 1.18.

And I found something in golang link, which create `pclntab`, in Function [`ld/pcln.go:generatePCHeader`](https://github.com/golang/go/blob/master/src/cmd/link/internal/ld/pcln.go#L230), we will see `textStartOff := int64(8 + 2*ctxt.Arch.PtrSize)` and set `ldr.Lookup("runtime.text", 0)` to this offset, this function will called by [`link.pclntab`](https://github.com/golang/go/blob/master/src/cmd/link/internal/ld/pcln.go#L757) that  initializes the pclntab symbol.
And we see [`generateFunctab`](https://github.com/golang/go/blob/master/src/cmd/link/internal/ld/pcln.go#L535), this function also called by `link.pclntab`, this function call function [`writePCToFunc`](https://github.com/golang/go/blob/master/src/cmd/link/internal/ld/pcln.go#L604) and [`writeFuncs`](https://github.com/golang/go/blob/master/src/cmd/link/internal/ld/pcln.go#L625).

This two function all use this code when set function offset to `pclntab`. 
```golang
textStart := ldr.SymValue(ldr.Lookup("runtime.text", 0))

off := ldr.SymValue(s) - textStart
```

and in `debug/gosym`, we'll see [this code](https://github.com/golang/go/blob/master/src/debug/gosym/pclntab.go#L416) when parse function offset and calculate function address from `pclntab`: 
```golang
	u := f.uint(f.functab[2*i*f.sz:])
	if f.version >= ver118 {
		u += f.textStart
	}
```

**Comment by sancppp** [comment]
I did a simple test of [v0.7.6](https://github.com/gojue/ecapture/tree/v0.7.6) and [PR#516](https://github.com/gojue/ecapture/pull/516) in an x64 Ubuntu 22.04 environment.

The test was to execute the command `sudo . /bin/ecapture gotls --elfpath=/usr/bin/dockerd --hex` while executing `docker login` in another terminal.

The result shows that eCapture was able to correctly find the crypto/tls.(*Conn).Read symbol entry, but was unable to locate the RET command:
image

 At the same time, `docker login` returns a connect error:
CleanShot_2024-04-08_at_22 34 06@2x

File: user/config/config_gotls.go (hunk 2 above)
Comment (cfc4n) [reply user/config/config_gotls.go:317]: Thank you very much for your explanation, I wasn't very familiar with this part before, I learned a lot.

**Review by cfc4n — APPROVED** [review]
LGTM, thanks.

**Comment by sancppp** [comment]
> I think the `connect error` issue probably has nothing to do with the hook position. The hook here is looking for the `RET` instruction to hook, so it shouldn't affect network communication. There must be another reason, maybe it's just a sporadic network failure.

It looks like eCpture is indeed affecting docker's network communication.

Normally docker login should return `Error response from daemon: Get "https://registry-1.docker.io/v2/": unauthorized: incorrect username or password`, however eCapture intervenes and returns an EOF error, while docker pull also returns an EOF error.

The test is relatively simple. I'm not sure if eCapture hooking other golang projects would cause the same thing.

CleanShot_2024-04-09_at_21 33 03

**Pull request merged**